examples/pytorch/bgnn/BGNN.py

```diff
@@ -0,0 +1,356 @@
+
+    def log_epoch(self, pbar, metrics, epoch, loss, epoch_time, logging_epochs, metric_name='loss'):
+        train_metric, val_metric, test_metric = metrics[metric_name][-1]
+        if epoch and epoch % logging_epochs == 0:
+            pbar.set_description(
+                "Epoch {:05d} | Loss {:.3f} | Loss {:.3f}/{:.3f}/{:.3f} | Time {:.4f}".format(epoch, loss,
+                                                                                              train_metric,
+                                                                                              val_metric,
+                                                                                              test_metric,
+                                                                                              epoch_time))
+
+    def fit(self, graph, X, y,
+            train_mask, val_mask, test_mask,
+            cat_features = None,
+            num_epochs=100,
+            patience=10,
+            logging_epochs=1,
+            metric_name='loss',
+            normalize_features=False,
+            replace_na=True,
+            ):
+        '''
+
+        :param graph: dgl graph
+        :param X: numpy matrix of input features. Can be composed of categorical features with missing values.
+        :param y: numpy matrix of target labels. Can be multidimensional targets.
+        :param train_mask: list of integers (rows in X that correspond to train set)
+        :param val_mask: list of integers (rows in X that correspond to valid set)
+        :param test_mask: list of integers (rows in X that correspond to test set)
+        :param cat_features: list of integers (columns in X that correspond to categorical features)
+        :param num_epochs: number of epochs to run
+        :param patience: number of epochs to wait until early stopping
+        :param logging_epochs: log every n epoch
+        :param metric_name: metric to use for early stopping
+        :param normalize_features: normalize original input features X (column wise)
+        :param replace_na: replace NA in X
+        :return: metrics evaluated during training
+        '''
+
+        # initialize for early stopping and metrics
+        if metric_name in ['r2', 'accuracy']:
+            best_metric = [np.float('-inf')] * 3  # for train/val/test
+        else:
+            best_metric = [np.float('inf')] * 3  # for train/val/test
+
+        best_val_epoch = 0
+        epochs_since_last_best_metric = 0
+        metrics = ddict(list)
+        if cat_features is None:
+            cat_features = []
+
+        if self.task == 'regression':
+            self.out_dim = y.shape[1]
+        elif self.task == 'classification':
+            self.out_dim = len(set(y.iloc[test_mask, 0]))
+        self.in_dim = self.out_dim + X.shape[1] if self.append_gbdt_pred else self.out_dim
+
+        gbdt_X_train = X.iloc[train_mask]
+        gbdt_y_train = y.iloc[train_mask]
+        gbdt_alpha = 1
```

Comment:
Is this a tunable hyperparameter? Consider expose it to the `__init__` function with default value.